cmd/storj-mount/mount.go

```diff
@@ -0,0 +1,561 @@
+}
+
+func (sf *storjFS) Create(name string, flags uint32, mode uint32, context *fuse.Context) (file nodefs.File, code fuse.Status) {
+	zap.S().Debug("Create: ", name)
+
+	return sf.addCreatedFile(name, newStorjFile(sf.ctx, name, sf.bucket, true, sf)), fuse.OK
+}
+
+func (sf *storjFS) addCreatedFile(name string, file *storjFile) *storjFile {
+	sf.createdFiles[name] = file
+	return file
+}
+
+func (sf *storjFS) removeCreatedFile(name string) {
+	delete(sf.createdFiles, name)
+}
+
+func (sf *storjFS) listObjects(ctx context.Context, name string, recursive bool, handler func([]storj.Object) error) error {
+	startAfter := ""
+
+	for {
+		list, err := sf.bucket.ListObjects(sf.ctx, &storj.ListOptions{
+			Direction: storj.After,
+			Cursor:    startAfter,
+			Prefix:    name,
+			Recursive: recursive,
+		})
+		if err != nil {
+			return err
+		}
+
+		err = handler(list.Items)
+		if err != nil {
+			return err
+		}
+
+		if !list.More {
+			break
+		}
+
+		startAfter = list.Items[len(list.Items)-1].Path
+	}
+
+	return nil
+}
+
+func (sf *storjFS) Unlink(name string, context *fuse.Context) (code fuse.Status) {
+	zap.S().Debug("Unlink: ", name)
+
+	err := sf.bucket.DeleteObject(sf.ctx, name)
+	if err != nil {
+		if storj.ErrObjectNotFound.Has(err) {
+			return fuse.ENOENT
+		}
+		return fuse.EIO
+	}
+
+	return fuse.OK
+}
+
+type storjFile struct {
+	ctx             context.Context
+	bucket          *libuplink.Bucket
+	created         bool
+	name            string
+	size            uint64
+	mtime           uint64
+	reader          io.ReadCloser
+	writer          io.WriteCloser
+	predictedOffset int64
+	FS              *storjFS
+
+	nodefs.File
+}
+
+func newStorjFile(ctx context.Context, name string, bucket *libuplink.Bucket, created bool, FS *storjFS) *storjFile {
+	return &storjFile{
+		name:    name,
+		ctx:     ctx,
+		bucket:  bucket,
+		mtime:   uint64(time.Now().Unix()),
+		created: created,
+		FS:      FS,
+		File:    nodefs.NewDefaultFile(),
+	}
+}
+
+func (f *storjFile) GetAttr(attr *fuse.Attr) fuse.Status {
+	zap.S().Debug("GetAttr file: ", f.name)
+
+	if f.created {
+		attr.Mode = fuse.S_IFREG | 0644
+		if f.size != 0 {
+			attr.Size = f.size
+		}
+		attr.Mtime = f.mtime
+		return fuse.OK
+	}
+	return fuse.ENOSYS
+}
+
+func (f *storjFile) Read(buf []byte, off int64) (res fuse.ReadResult, code fuse.Status) {
+	// Detect if offset was moved manually (e.g. stream rev/fwd)
+	if off != f.predictedOffset {
+		f.closeReader()
+	}
+
+	reader, err := f.getReader(off)
+	if err != nil {
+		if storj.ErrObjectNotFound.Has(err) {
+			return nil, fuse.ENOENT
+		}
+		return nil, fuse.EIO
+	}
+
+	n, err := io.ReadFull(reader, buf)
+	if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+		return nil, fuse.EIO
+	}
+
+	f.predictedOffset = off + int64(n)
+
+	return fuse.ReadResultData(buf[:n]), fuse.OK
+}
+
+func (f *storjFile) Write(data []byte, off int64) (uint32, fuse.Status) {
+	writer, err := f.getWriter(off)
```

Comment:
offset is not always ordered from 0. Writes might come in in the order "100, 500, 150, 0, 200" if the writing application is seeking around. The previous version of this had an expectedOffset for both reads and writes that was compared. We should do that here (or in getWriter) and throw an error if the offset does not match the expected offset